File: typescript/lib/base/models/circle.ts
import { BaseShape, Dimensional, Displayable } from "../commons";

@Displayable
export class Circle extends BaseShape implements Dimensional {
    private readonly radius: number;
    private _color: string;

    constructor(radius: number) {
        const dim1: number = radius > 0 ? radius : 1;

        super([
            { radius: Math.max(radius, 1) },
            { diameter: Math.max(radius, 1) },
            { curvature: 1 / Math.max(radius, 1) },
        ]);

        this.radius = dim1;
        this._color = "None";
    }

    get color(): string {
        return this._color;
    }

    set color(value: string) {
        this._color = value;
    }

    printDescription(): void {
        console.log(`Circle radius: ${this.radius}`);
        console.log(`Circle dimensions: ${this.formatDimensions()}`);
        console.log(`Circle color: ${this.color}`);
        console.log(`Circle area: ${this.area()}`);
        console.log(`Circle perimeter: ${this.perimeter()}`);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        console.log(`Circle is displayable: ${(this as any).shouldDisplay()}`);
        console.log(`Circle is a shape: ${Circle.validateShape()}`);
    }

    area(): number {
        return Math.PI * Math.pow(this.radius, 2);
    }

    perimeter(): number {
        return 2 * Math.PI * this.radius;
    }

    private formatDimensions(): string {
        const parts: string[] = this.dimensions.map((dim) => {
            const [key, value] = Object.entries(dim)[0] ?? ["", 0];
            return `${key}: ${value.toFixed(3)}`;
        });

        return parts.join(" | ");
    }
}

File: dart/lib/src/base/models/circle.dart
import "dart:math";

import "package:shapes/base/commons.dart";

final class Circle extends BaseShape with Displayable implements Dimensional {
  double? _radius;
  String? _color;

  Circle(double radius)
    : super([
        {"radius": max(radius, 1)},
        {"diameter": max(radius, 1) * 2},
        {"curvature": 1 / max(radius, 1)},
      ]) {
    double dim1 = radius > 0 ? radius : 1;

    _radius = dim1;
    _color = "None";
  }

  String get color {
    return _color!;
  }

  set color(String value) {
    _color = value;
  }

  @override
  void printDescription() {
    print("Circle radius: $_radius");
    print("Circle dimensions: ${_formatDimensions()}");
    print("Circle color: $color");
    print("Circle area: ${area()}");
    print("Circle perimeter: ${perimeter()}");
    print("Circle is displayable: ${shouldDisplay()}");
    print("Circle is a shape: ${BaseShape.validateShape()}");
  }

  @override
  double area() {
    return pi * pow(_radius!, 2);
  }

  @override
  double perimeter() {
    return 2 * pi * _radius!;
  }

  String _formatDimensions() {
    List<String> parts =
        dimensions.map((dimension) {
          return dimension.entries
              .map((e) {
                return "${e.key}: ${e.value.toStringAsFixed(3)}";
              })
              .join(", ");
        }).toList();

    return parts.join(" | ");
  }
}
